src/service-worker/core.js

```diff
@@ -0,0 +1,299 @@
+      }
+
+      const files = transaction.objectStore('js-files');
+      const versions = files.index('versions');
+
+      // Do not serve versions older than two weeks.
+      const cutoff = Number(RELEASE_DATE.setDate(-14));
+      const range = IDBKeyRange.upperBound(cutoff);
+
+      // We need to find every file that has an old version, then prune that
+      // version from the db and cache.
+      cacheCleanup = Promise.resolve(versions.openCursor(range).while(cur => {
+        const item = cur.value;
+        const removal = {
+          url: item.url,
+          versions: item.versions.filter(v => (v <= cutoff))
+        };
+
+        // Remove old versions from our db.
+        item.versions = item.versions.filter(v => v > cutoff);
+        return cur.put(item).then(() => removal);
+      })).then(removals => {
+        // Prune all versions of all files from the cache.
+        const deletes = removals.map(removal => {
+          // Prune all versions of this file from the cache
+          const deletes = removal.versions.map(version => {
+            const url = versionedUrl(removal.url, version);
+            return cache.delete(url);
+          });
+
+          return Promise.all(deletes);
+        });
+
+        return Promise.all(deletes);
+      });
+    }
+  });
+}).then(result => {
+  db = result;
+});
+
+self.addEventListener('install', function(install) {
+  install.waitUntil(Promise.all([cachePromise, cacheCleanup, dbPromise]));
+});
+
+// Setup the Fetch listener
+// My assumptions:
+//   - Doc requests one uniform AMP release version for all files, anything
+//     else is malarkey.
+self.addEventListener('fetch', function(event) {
+  const request = event.request;
+  const clientId = event.clientId;
+  const url = request.url;
+  let response;
+
+  // We only cache CDN JS files, and we need a clientId to do our magic.
+  if (clientId && isCdnJsFile(url)) {
+    const requestFile = basename(url);
+    const requestVersion = ampVersion(url);
+
+    // What version do we have for this client?
+    response = Promise.resolve(clientsMap[clientId]).then(version => {
+      // If we already registered this client, we must always use the same
+      // version.
+      if (version) {
+        return version;
+      }
+
+      // If not, do we have this version cached, if so serve it.
+      // do we have a newer version cached? Check the db, if so serve it.
+      // If not, get the newest version, cache it, serve it.
+      return db.transaction('js-files', 'readonly').run(transaction => {
+        const files = transaction.objectStore('js-files');
+        return files.get(requestFile);
+      }).then((item = {}) => {
+        const versions = item.versions;
+        if (!versions || versions.length === 0) {
+          return VERSION;
+        }
+
+        const isVersionCached = versions.indexOf(requestVersion) > -1;
+        // TODO Make this smarter. We should probably select the version
```

Comment:
This will always use a uniform version for a single document due to https://github.com/jridgewell/amphtml/blob/03466c7810dac4594c51c7af988b008f9eda456a/src/service-worker/core.js#L218-L223. This codepath will only be executed for the first JS file requested from the page, everything else will use the version we determined (because it'll be stored in `clientsMap`).